bots/openprescribing/openprescribing.py

```diff
@@ -0,0 +1,192 @@
+    if deploy_in_progress():
+        message.reply(
+            "Deploy underway. Will start another when it's finished",
+            in_thread=True
+        )
+    else:
+        if message.body['ts']:
+            msg = "Deploying in {} seconds".format(DEPLOY_DELAY)
+            message.reply(msg, in_thread=True)
+        else:
+            msg = "PR merged. "
+            msg += "Deploying in {} seconds".format(DEPLOY_DELAY)
+            # This was triggered from github webhooks - no thread
+            message.send_webapi(msg)
+    reset_or_deploy_timer(DEPLOY_DELAY, message)
+
+
+@respond_to('op deploy now', re.IGNORECASE)
+@suppressed
+def deploy_live_now(message):
+    message.reply(
+        "Deploying now".format(DEPLOY_DELAY),
+        in_thread=True)
+    reset_or_deploy_timer(0, message)
+
+
+@respond_to('op cancel deploy', re.IGNORECASE)
+@suppressed
+def cancel_deploy_live(message):
+    reset_or_deploy_timer(None, message)
+    message.reply("Cancelled", in_thread=True)
+
+
+@respond_to(r'op clear cache', re.IGNORECASE)
+def clear_cache(message):
+    result = safe_execute(clear_cloudflare, hosts=HOSTS)
+    message.reply("Cache cleared:\n\n {}".format(result), in_thread=True)
+
+
+@respond_to('op cancel suppression', re.IGNORECASE)
+def cancel_suppression(message):
+    if flags.deploy_suppressed:
+        flags.deploy_suppressed = None
+        message.reply("Cancelled", in_thread=True)
+    else:
+        message.reply("No suppressions to cancel", in_thread=True)
+
+
+@respond_to('op status', re.IGNORECASE)
+def show_status(message):
+    msgs = []
+    if flags.deploy_suppressed:
+        start_time, end_time = flags.deploy_suppressed
+        msgs.append("Deploys suppressed from {} to {}`".format(
+            start_time.strftime(TIME_FMT),
+            end_time.strftime(TIME_FMT)))
+    else:
+        if flags.deploy_countdown is None:
+            msgs.append("No deploys in progress")
+        elif flags.deploy_countdown < 1:
+            msgs.append("Deploy in progress")
+        else:
+            msgs.append(
+                "Deploy due in {} seconds".format(flags.deploy_countdown)),
+    message.reply("\n".join(msgs), in_thread=True)
+
+
+@respond_to(r'op suppress from (.*) to (.*)', re.IGNORECASE)
+def suppress_deploy(message, start_time, end_time):
+    start_time = parse(start_time)
+    end_time = parse(end_time)
+    flags.deploy_suppressed = [start_time, end_time]
+    message.reply(
+        "Deployment suppressed from {} until {}. "
+        "Cancel with `cancel suppression`".format(
+            start_time.strftime(TIME_FMT),
+            end_time.strftime(TIME_FMT)),
+        in_thread=True
+    )
+    if flags.deploy_countdown:
+        deploy_due_at = datetime.now() + flags.deploy_countdown
+        if deploy_due_at <= end_time:
+            reset_or_deploy_timer(None, message)
+            message.reply("Current deployment cancelled", in_thread=True)
+
+
+def deploy_timer(message):
```

Comment:
This code is not obviously correct, and and I'm not convinced that there aren't race conditions.

I think it'd be clearer and safer if there was a single thread, always running, that watched the clock, and rather than having a countdown, this thread compared the current time with a global `deploy_time`, which gets set or cleared by the appropriate commands.

The thread function might look something like this:

```
def deploy_timer():
    while True:
        sleep(1)

        if not flags.deploy_time:
            # no deploy scheduled
            continue

        if time.now() < flags.deploy_time:
            # deploy is scheduled in the future
            continue

        if flags.deploy_suppressed:
            # deploy has been suppressed

            start_time, end_time = flags.deploy_suppressed
            if start_time < time.now() < end_time:
                # deploy is currently suppressed
                continue
            elif end_time <= time.now():
                # suppression was in past, so clear it
                flags.deploy_suppressed = None
            else:
                # suppression is in future, so do nothing
                pass

        # we are able to deploy!
        flags.deploy_in_progress = True
        safe_execute(...)
        flags.deploy_in_progress = False

        flags.deploy_message.reply(...)
```

Reply:
I agree this is clearer and therefore probably safer.

The down side is that having a single thread means there is no context for knowing which was the calling command.

I've come to the view (per #8) that threading our bot responses is not helpful, so perhaps this context is no longer helpful either. If we ever wanted to reintroduce threading, then we'd want some kind of per-deploy-message threading again.

So while I agree the proposed code is better I'm slightly loathe to change it right now in case we decide message threading would be useful (or otherwise want to refer to the originating message) and forget that it had previously been implemented in this way.  Does that make sense, or am I unnecessarily keeping code without a clear use case in mind?